test/storage-luatest/storage_1_1_test.lua

```diff
@@ -157,3 +157,99 @@ test_group.test_bucket_send_field_types = function(g)
         box.space.test:drop()
     end)
 end
+
+--
+-- Custom triggers on bucket changes
+--
+test_group.test_on_bucket_event = function(g)
+    -- Initial problem: buckets migration looks like a regular
+    -- client data activity, indistinguishable in the replication flow.
+    -- In order to allow efficient change data capturing we can add a fake
+    -- operation in the beginning of the transaction, so that a client could
+    -- determing which operations can be skipped (using HEADER.tsn) as they
+    -- were done not by the user but by gc or rebalancer.
+    vtest.cluster_exec_each_master(g, function(engine)
+        -- Space, in which these fake operations will be written
+        rawset(_G, 'notify_name', '_vshard_migration_notify_' .. engine)
+        box.schema.space.create(_G.notify_name, {
+            engine = engine,
+        }):create_index('primary')
+        local s = box.schema.space.create('data', {
+             engine = engine,
+             format = {
+                {'id', 'unsigned'},
+                {'bucket_id', 'unsigned'},
+             },
+        })
+        s:create_index('id', {parts = {'id'}})
+        s:create_index('bucket_id', {parts = {'bucket_id'}, unique = false})
+        rawset(_G, 'notify_trigger', function(event_type, bucket_id)
+            if event_type == 'bucket_data_recv_txn' or
+                    event_type == 'bucket_data_gc_txn' then
+                box.space[_G.notify_name]:delete({bucket_id})
+            end
+        end)
+        -- Set trigger
+        ivshard.storage.on_bucket_event(_G.notify_trigger)
+    end, {g.params.engine})
+
+    local uuid_1 = g.replica_1_a:exec(function()
+        return box.info.cluster.uuid
+    end)
+
+    g.replica_2_a:exec(function(uuid_1)
+        local bid = _G.get_first_bucket()
+        box.space.data:insert({bid, bid})
+        ivshard.storage.bucket_send(bid, uuid_1)
+    end, {uuid_1})
+
+    -- Read xlog, get all messages written to the special space
+    local function get_notify_wal_rows(storage)
+        return storage:exec(function()
+            local xlog = require('xlog')
+            local notifications = {}
+            local notify_space = box.space[_G.notify_name]
+            for _, request in xlog.pairs('00000000000000000000.xlog') do
```

Comment:
That for sure is quite a complicated way. Lets not depend on whether it is stored in xlog or not, and for sure not depend on file names. To validate that everything ends up in one txn you can use `box.txn.on_commit()` trigger. Here you can read up all the rows and append them to some global table. Then further down the test you take that table and check each saved txn content. That would allow not to touch the files and nor look for what is the current's xlog name. And you won't depend on 1.10 - txn triggers are supported there.

Reply:
Such a great proposal. Fixed